File: classical_clang/quick_sort/quick_sort.c
/*
 * =====================================================================================
 *
 *       Filename:  quick_sort.c
 *
 *    Description:  快排
 *
 *        Version:  1.0
 *        Created:  2019-07-22
 *       Revision:  none
 *       Compiler:  gcc
 *
 *         Author:  fangyuan
 *
 * =====================================================================================
 */

#include <stdio.h>
#include <stdlib.h>

void swap(int *a, int *b) {
        int temp;

        temp = *a;
        *a = *b;
        *b = temp;

        return;
}

void display(int *array) {
        int i;
        for (i = 0; i < BUFSIZ; i++)
                if (array[i])
                        printf("%-3d", array[i]);

        printf("\n");
}

void quick_sort(int *array, int begin, int end) {
        int i, j;
        if (begin < end) {
                i = begin + 1;
                j = end;
                while (i < j) {
                        if (array[i] > array[begin]) {
                                swap(&array[i], &array[j]);
                                j--;
                        } else {
                                i++;
                        }
                }

                if (array[i] >= array[begin])
                        i--;

                swap(&array[begin], &array[i]);
                quick_sort(array, begin, i);
                quick_sort(array, j, end);
        }
}

int main() {
        int array[BUFSIZ] = {5, 2, 3, 1};
        quick_sort(array, 0, 3);
        display(array);

        return 0;
}

File: classical_java/quick_sort/QuickSort.java
public class QuickSort {
        public static void main(String[] args) {
                int[] nums;
                int l;

                nums = new int[] {12, 23, 4, 65, 76, 22, 13, 2, 1, 0};
                l = nums.length;

                quick_sort(nums, 0, l - 1);
                view_list(nums);
                return;
        }
        private static void quick_sort(int[] nums, int begin, int end) {

                int i, j, tmp;

                if (begin < end) {
                        i = begin + 1;
                        j = end;
                        for (; i < j;) {
                                if (nums[i] > nums[begin]) {
                                        // swap
                                        tmp = nums[i];
                                        nums[i] = nums[j];
                                        nums[j] = tmp;
                                        j--;
                                } else {
                                        i++;
                                }
                        }
                        // j == i
                        if (nums[i] >= nums[begin]) {
                                i--;
                        }
                        tmp = nums[i];
                        nums[i] = nums[begin];
                        nums[begin] = tmp;

                        quick_sort(nums, begin, i);
                        quick_sort(nums, j, end);
                }
                return;
        }
        private static void view_list(int[] nums) {
                int i, l;
                l = nums.length;
                for (i = 0; i < l; i++) {
                        System.out.printf("%d ", nums[i]);
                }
                System.out.print("\n");
        }
}
